app/src/main/java/com/example/pix/home/adapters/SearchAdapter.java

```diff
@@ -0,0 +1,147 @@
+            this.tvName = itemView.findViewById(R.id.chats_name);
+            this.tvStatus = itemView.findViewById(R.id.chats_status);
+            this.tvTime = itemView.findViewById(R.id.chats_time);
+            this.tvPix = itemView.findViewById(R.id.chats_pix);
+            this.llSelect = itemView.findViewById(R.id.chats_select);
+        }
+
+        public void bind(ParseUser user) {
+            this.tvPix.setText("" + user.getInt("pix") + "P");
+            this.llSelect.setOnClickListener(view -> {
+                // If we select a result, we will want to go to ChatActivity
+                Intent i = new Intent(context, ChatActivity.class);
+
+                // Look to see if a Chat with the selected User already exists
+                Chat chat = findChat(true, user);
+
+                // If we indicated that this Adapter was made with intent to send pic,
+                if (newPic)
+                    i.putExtra("newPic", newPic);
+
+                // If we couldnt find an existing CHat, make a new one
+                if (chat == null) {
+                    Chat newChat = new Chat();
+                    newChat.setUser(ParseUser.getCurrentUser());
+                    newChat.setFriend(user);
+                    newChat.saveInBackground(e -> {
+                        i.putExtra("chat", newChat.getObjectId());
+                        context.startActivity(i);
+                    });
+                    return;
+                }
+                i.putExtra("chat", chat.getObjectId());
+                context.startActivity(i);
+            });
+            ParseFile pic = user.getParseFile("profile");
+            if (pic != null)
+                Glide.with(context)
+                        .load(pic
+                                .getUrl())
+                        .circleCrop()
+                        .into(this.ivProfile);
+            this.tvName.setText("" + user.getUsername());
+            this.tvStatus.setVisibility(View.GONE);
+            this.tvTime.setVisibility(View.GONE);
+        }
+    }
+
+    // Note: userOne and userTwo in Chat table is both people in the chat in no particular order,
+    // this we need to check both cases:
+    // User is userOne and friend is userTwo
+    // friend is userOne and user is userTwo
+    private Chat findChat(boolean userOne, ParseUser user) {
+        ParseQuery<Chat> q = ParseQuery.getQuery(Chat.class);
+        q.include("userOne");
+        q.include("userTwo");
+        if (userOne) {
+            q.whereEqualTo("userOne", ParseUser.getCurrentUser());
+            q.whereEqualTo("userTwo", user);
+        } else {
+            q.whereEqualTo("userOne", user);
+            q.whereEqualTo("userTwo", ParseUser.getCurrentUser());
+        }
+        try {
+            Chat res = q.getFirst();
+            if (res != null)
+                return res;
+            if (userOne)
+                return findChat(false, user);
```

Comment:
Also, I find the "userOne" and "userTwo" relationship a little confusing. Is it meat to represent the "sender" and "receiver" relationship? If so, you should rename your columns, if that's possible. If not, how do you know who is the sender or receiver of a message, or do you not need that information?
It would become useful if you decide to implement the feature that shows the sender's messages on one side, and receiver's on the other, which many messaging apps do.

Reply:
So the way my Parse db is set up, userOne and userTwo are the two Users who are a part of a particular Chat, in no order. This method, "findChat" needs to figure out if there exists a Chat involving a specific User "user" (the friend) and the current User. The way I did it is by creating a boolean userOne to tell us to first look for a Chat that has the current User as userOne and the given user(the friend) as userTwo. The if statement does this. If we find that there isnt a Chat like this, I recursed, but this time checking for the friend as userOne and current User as userTwo, if this also finds nothing, I returned null. I thought of writing the two possibilities as 2 different queries back to back instead of recursing, but I figured it would save a bit of code if I did it this way. Let me know if it is kind weird and I can try to figure out another way.

Reply:
It's fine if you want to use recursion to save code. I had a typo in the original comment, but I was asking, what if the caller calls the function like this: `findChat(false, ...)`, then the function won't return all the search result because it doesn't swap the users.
Think about what you want to achieve with this method, you just want to find all the chats the current user has with the other user, then the signature of the method should just be:
`private Chat findChat(ParseUser friend);`, without requiring the caller to know the implementation detail of the function (whether the current user is 'userOne' or 'userTwo').
To reuse your code here, the implementation will probably be:
```
private Chat findChatHelper(boolean userOne, ParseUser user) {
// whatever you currently have
}

private Chat findChat(ParseUser friend) {
   return findChatHelper(true, friend);
}
```

Reply:
Another way to implement this that might be a little cleaner:
```
private Chat findChatHelper(ParseUser user1, ParseUser user2) {...}

private Chat findChat(ParseUser friend) {
  Chat result = findChatHelper(ParseUser.getCurrentUser(), friend);
  if (result != null) {
    return result;
  }
  return findChatHelper(friend, ParseUser.getCurrentUser());
}
```
Without all the error handling. But you probably get the idea.

Also, I might have misunderstood what a "chat" means previously. It means a chat thread, but not a message right? Then I get why there's no order with 'userOne' and 'userTwo' now.

Reply:
Yes its a chat thread, and I'll clean the method up now. Thanks